pkg/solana/chainreader/event_read_binding.go

```diff
@@ -0,0 +1,245 @@
+package chainreader
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/gagliardetto/solana-go"
+
+	commoncodec "github.com/smartcontractkit/chainlink-common/pkg/codec"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/codec"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/logpoller"
+)
+
+type eventReadBinding struct {
+	namespace, genericName string
+	codec                  types.RemoteCodec
+	modifier               commoncodec.Modifier
+	key                    solana.PublicKey
+	remapper               remapHelper
+	indexedSubkeys         map[string]string
+	events                 EventsReader
+	eventSig               [logpoller.EventSignatureLength]byte
+}
+
+func newEventReadBinding(
+	namespace, genericName string,
+	indexedSubkeys map[string]string,
+	events EventsReader,
+	eventSig [logpoller.EventSignatureLength]byte,
+) *eventReadBinding {
```

Comment:
Assuming the ordered slice of onChainNames is passed into this constructor, we'd build the `eventReadBinding.indexedSubKeys` map like this:
```
subKeyIndexes := make(map[string]int, len(onChainNames))
for i, onChainName := range onChainNames {
     subKeyIndexes[onChainName] = i
}
```